probe/plugins/registry.go

```diff
@@ -0,0 +1,293 @@
+package plugins
+
+import (
+	"fmt"
+	"net/http"
+	"net/url"
+	"path/filepath"
+	"sort"
+	"strings"
+	"sync"
+	"syscall"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/ugorji/go/codec"
+	"golang.org/x/net/context"
+	"golang.org/x/net/context/ctxhttp"
+
+	"github.com/weaveworks/scope/common/backoff"
+	"github.com/weaveworks/scope/common/fs"
+	"github.com/weaveworks/scope/common/xfer"
+	"github.com/weaveworks/scope/report"
+)
+
+// Exposed for testing
+var (
+	transport = makeUnixRoundTripper
+)
+
+const (
+	pluginTimeout   = 500 * time.Millisecond
+	pluginRetry     = 5 * time.Second
+	pollingInterval = 5 * time.Second
+)
+
+// Registry maintains a list of available plugins by name.
+type Registry struct {
+	rootPath          string
+	apiVersion        string
+	handshakeMetadata map[string]string
+	pluginsBySocket   map[string]*Plugin
+	lock              sync.RWMutex
+	context           context.Context
+	cancel            context.CancelFunc
+}
+
+// NewRegistry creates a new registry which watches the given dir root for new
+// plugins, and adds them.
+func NewRegistry(rootPath, apiVersion string, handshakeMetadata map[string]string) (*Registry, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	r := &Registry{
+		rootPath:          rootPath,
+		apiVersion:        apiVersion,
+		handshakeMetadata: handshakeMetadata,
+		pluginsBySocket:   map[string]*Plugin{},
+		context:           ctx,
+		cancel:            cancel,
+	}
+	if err := r.scan(); err != nil {
+		r.Close()
+		return nil, err
+	}
+	go r.loop()
+	return r, nil
+}
+
+// loop periodically rescans for plugins
+func (r *Registry) loop() {
+	ticker := time.NewTicker(pollingInterval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-r.context.Done():
+			return
+		case <-ticker.C:
+			log.Debugf("plugins: scanning...")
+			if err := r.scan(); err != nil {
+				log.Warningf("plugins: error: %v", err)
+			}
+		}
+	}
+}
+
+// Rescan the plugins directory, load new plugins, and remove missing plugins
+func (r *Registry) scan() error {
+	sockets, err := r.sockets(r.rootPath)
+	if err != nil {
+		return err
+	}
+
+	r.lock.Lock()
+	plugins := map[string]*Plugin{}
+	// add (or keep) plugins which were found
+	for _, path := range sockets {
+		if plugin, ok := r.pluginsBySocket[path]; ok {
+			plugins[path] = plugin
+			continue
+		}
+		tr, err := transport(path, pluginTimeout)
+		if err != nil {
+			log.Warningf("plugins: error loading plugin %s: %v", path, err)
+			continue
+		}
+		client := &http.Client{Transport: tr, Timeout: pluginTimeout}
+		plugins[path] = NewPlugin(r.context, path, client, r.apiVersion, r.handshakeMetadata)
+	}
+	// remove plugins which weren't found
+	for path, plugin := range r.pluginsBySocket {
```

Comment:
Due to polling I don't think this will catch things such as a plugin removing the unix socket and re-recreating it between calls to `scan()`.

To solve this I think we should automatically de-register plugins when `GET /report` causes irrecoverable errors (e.g. connection errors).